### DISCUSSION TITLE
Resizable Chat Window

### DISCUSSION THREAD

**Discussion by KieranConlon** [intitni/CopilotForXcode#174] [Ideas]
I'm not sure if a resizable chat window is possible, but I feel it might be helpful in some circumstances.

Working on refactoring some code today, and the AI-suggested code in the chat window was wrapping a lot.  I feel that if I could resize the window to be a little wider then the code would be a little easier to read.

<img width="1800" alt="image" src="https://user-images.githubusercontent.com/33759139/234398826-93219348-d9b0-47ee-84f6-70852d48c590.png">

**Comment by intitni** [comment]
Maybe I should just detach the chat panel from the circular widget and place it in a separate window that can be moved around and resized, and even stay displayed when Xcode is inactive. What do you think?

**Reply by DatZig** [reply]
yes this would be Awesome! it would be super helpful to be able to see the code suggestions coming back from openai in a real large window so I can spot check what it is suggesting before I accept it ...

**Comment by KieranConlon** [comment]
I like the idea of having a detachable chat panel; that would be nice.
How difficult would it be to detach/undock chat to a resizable window and then attach/dock chat to the widget?

Having the chat panel attached to the widget works well in most situations.  
But some chats with AI can generate lengthy responses that scroll off the visible area in the 'docked' chat panel; in this scenario ', undocking' the chat to allow re-positioning and resizing the window would work.

**Reply by intitni** [reply]
I already have a working implementation. Detaching can now be done in the contextual menu of the widget and re-attaching can be done as presented in the video. 

https://user-images.githubusercontent.com/793147/234487157-507fdee1-878a-4100-9e75-4536c62c2387.mov

**Reply by KieranConlon** [reply]
That looks great to me.


I have another question/suggestion about font sizes - maybe I should start a separate ideas discussion for that.

**Comment by intitni** [comment]
A beta version has been release for this feature https://github.com/intitni/CopilotForXcode/releases/tag/0.14.0.beta.1

**Reply by KieranConlon** [reply]
v14.0.b.1 installed. Detachable chat working great. :)

Just to note...
I did notice on first run of this version, the app seemed very slow to start and all of the button controls on the main app window were not selectable for maybe 60-90 seconds.  I was trying to click "Reload Launch Agent", to make sure the service was running.
Toggles and drop-downs were working OK.

Anyway, after about 90 sec everything was working as expected.

**Reply by intitni** [reply]
I also had the same experience where the issue occurred only once and never again. Since only buttons were not working, I would just blame SwiftUI.

**Comment by shanecowherd** [comment]
Thank you so much, this is exactly what I wanted, I've been using MacGPT on the side for this exact use case.

Well done!

**Comment by intitni** [comment]
Beta 2 released that fixed the app freezing issue https://github.com/intitni/CopilotForXcode/releases/tag/0.14.0.beta.2

**Comment by intitni** [comment]
[Beta 4 is out](https://github.com/intitni/CopilotForXcode/releases/tag/0.14.0.beta.4)

**Comment by DatZig** [comment]
the resizable chat window is awesome. I do have the occasional lag when I copy a big chunk of code into it and also type in there. Also, sometimes when I'm typing and press shift+enter to go to a new line it sends the incomplete message to gpt ...

**Reply by intitni** [reply]
Could you please submit a bug report in issues so that I can remember to look into it later? I suspect that the SwiftUI TextField may not be able to handle large content efficiently. Also, about the shift+enter issue, I haven't experienced it before. Is it possible to reproduce it by following certain steps?
